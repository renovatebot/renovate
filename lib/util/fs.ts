import * as fs from 'fs-extra';
import { join, parse } from 'upath';
import { RenovateConfig } from '../config/common';
import { logger } from '../logger';

let localDir = '';

export function setFsConfig(config: Partial<RenovateConfig>): void {
  localDir = config.localDir;
}

export function getSubDirectory(fileName: string): string {
  return parse(fileName).dir;
}

export function getSiblingFileName(
  existingFileNameWithPath: string,
  otherFileName: string
): string {
  const subDirectory = getSubDirectory(existingFileNameWithPath);
  return join(subDirectory, otherFileName);
}

export async function readLocalFile(fileName: string): Promise<Buffer>;
export async function readLocalFile(
  fileName: string,
  encoding: 'utf8'
): Promise<string>;
export async function readLocalFile(
  fileName: string,
  encoding?: string
): Promise<string | Buffer> {
  const localFileName = join(localDir, fileName);
  try {
    const fileContent = await fs.readFile(localFileName, encoding);
    return fileContent;
  } catch (err) {
    logger.trace({ err }, 'Error reading local file');
    return null;
  }
}

export async function writeLocalFile(
  fileName: string,
  fileContent: string
): Promise<void> {
  const localFileName = join(localDir, fileName);
  await fs.outputFile(localFileName, fileContent);
}

export async function deleteLocalFile(fileName: string): Promise<void> {
  await fs.remove(fileName);
}

// istanbul ignore next
export async function ensureDir(dirName): Promise<void> {
  await fs.ensureDir(dirName);
}

// istanbul ignore next
export async function ensureLocalDir(dirName): Promise<void> {
  const localDirName = join(localDir, dirName);
  await fs.ensureDir(localDirName);
}
