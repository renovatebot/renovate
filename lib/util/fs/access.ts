import upath from 'upath';
import { FILE_ACCESS_VIOLATION_ERROR } from '../../constants/error-messages';
import { logger } from '../../logger';

export function assertBaseDir(path: string, baseDir: string): void {
  if (!path.startsWith(upath.resolve(baseDir))) {
    logger.warn(
      { path, baseDir },
      'Preventing access to file outside the base directory'
    );
    throw new Error(FILE_ACCESS_VIOLATION_ERROR);
  }
}
