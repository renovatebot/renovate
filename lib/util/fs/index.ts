import * as fs from 'fs-extra';
import { join, parse } from 'upath';
import { RenovateConfig } from '../../config/common';
import { logger } from '../../logger';

export * from './proxies';

let localDir = '';
let cacheDir = '';

export function setFsConfig(config: Partial<RenovateConfig>): void {
  localDir = config.localDir;
  cacheDir = config.cacheDir;
}

export function getSubDirectory(fileName: string): string {
  return parse(fileName).dir;
}

export function getSiblingFileName(
  existingFileNameWithPath: string,
  otherFileName: string
): string {
  const subDirectory = getSubDirectory(existingFileNameWithPath);
  return join(subDirectory, otherFileName);
}

export async function readLocalFile(fileName: string): Promise<Buffer>;
export async function readLocalFile(
  fileName: string,
  encoding: 'utf8'
): Promise<string>;
export async function readLocalFile(
  fileName: string,
  encoding?: string
): Promise<string | Buffer> {
  const localFileName = join(localDir, fileName);
  try {
    const fileContent = await fs.readFile(localFileName, encoding);
    return fileContent;
  } catch (err) {
    logger.trace({ err }, 'Error reading local file');
    return null;
  }
}

export async function writeLocalFile(
  fileName: string,
  fileContent: string
): Promise<void> {
  const localFileName = join(localDir, fileName);
  await fs.outputFile(localFileName, fileContent);
}

export async function deleteLocalFile(fileName: string): Promise<void> {
  if (localDir) {
    const localFileName = join(localDir, fileName);
    await fs.remove(localFileName);
  }
}

// istanbul ignore next
export async function ensureDir(dirName: string): Promise<void> {
  await fs.ensureDir(dirName);
}

// istanbul ignore next
export async function ensureLocalDir(dirName: string): Promise<void> {
  const localDirName = join(localDir, dirName);
  await fs.ensureDir(localDirName);
}

export async function ensureCacheDir(
  dirName: string,
  envPathVar?: string
): Promise<string> {
  const envCacheDirName = envPathVar ? process.env[envPathVar] : null;
  const cacheDirName = envCacheDirName || join(cacheDir, dirName);
  await fs.ensureDir(cacheDirName);
  return cacheDirName;
}

/**
 * Return the path of the private cache directory. This directory is wiped
 * between repositories, so they can be used to store private registries' index
 * without risk of that information leaking to other repositories/users.
 */
export function privateCacheDir(): string {
  return join(cacheDir, '__renovate-private-cache');
}

export function localPathExists(pathName: string): Promise<boolean> {
  // Works for both files as well as directories
  return fs
    .stat(join(localDir, pathName))
    .then((s) => !!s)
    .catch(() => false);
}
