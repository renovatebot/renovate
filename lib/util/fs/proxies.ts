import * as fs from 'fs-extra';
import { MoveOptions, WriteFileOptions } from 'fs-extra';

// istanbul ignore next
export function stat(path: string | Buffer): Promise<fs.Stats> {
  return fs.stat(path);
}

// istanbul ignore next
export function chmod(
  path: string | Buffer,
  mode: string | number
): Promise<void> {
  return fs.chmod(path, mode);
}

export async function readFile(fileName: string): Promise<Buffer>;
export async function readFile(
  fileName: string,
  encoding: 'utf8'
): Promise<string>;
export function readFile(
  fileName: string,
  encoding?: string
): Promise<string | Buffer> {
  return fs.readFile(fileName, encoding);
}

// istanbul ignore next
export function writeFile(
  fileName: string,
  fileContent: string
): Promise<void> {
  return fs.writeFile(fileName, fileContent);
}

// istanbul ignore next
export function outputFile(
  file: string,
  data: any,
  options?: WriteFileOptions | string
): Promise<void> {
  return fs.outputFile(file, data, options ?? {});
}

export function remove(dir: string): Promise<void> {
  return fs.remove(dir);
}

// istanbul ignore next
export function unlink(path: string | Buffer): Promise<void> {
  return fs.unlink(path);
}

// istanbul ignore next
export function exists(path: string): Promise<boolean> {
  return fs.pathExists(path);
}

// istanbul ignore next
export function pathExists(path: string): Promise<boolean> {
  return fs.pathExists(path);
}

// istanbul ignore next
export function move(
  src: string,
  dest: string,
  options?: MoveOptions
): Promise<void> {
  return fs.move(src, dest, options ?? {});
}
