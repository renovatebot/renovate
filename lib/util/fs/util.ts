import upath from 'upath';
import { GlobalConfig } from '../../config/global';
import { FILE_ACCESS_VIOLATION_ERROR } from '../../constants/error-messages';
import { logger } from '../../logger';

function assertBaseDir(path: string, baseDir: string): void {
  if (!path.startsWith(upath.resolve(baseDir))) {
    logger.debug(
      { path, baseDir },
      'Preventing access to file outside the base directory'
    );
    throw new Error(FILE_ACCESS_VIOLATION_ERROR);
  }
}

function ensurePath(path: string, key: 'localDir' | 'cacheDir'): string {
  const baseDir = upath.resolve(GlobalConfig.get(key)!);
  const fullPath = upath.resolve(
    upath.isAbsolute(path) ? path : upath.join(baseDir, path)
  );
  assertBaseDir(fullPath, baseDir);
  return fullPath;
}

export function ensureLocalPath(path: string): string {
  return ensurePath(path, 'localDir');
}

export function ensureCachePath(path: string): string {
  return ensurePath(path, 'cacheDir');
}
