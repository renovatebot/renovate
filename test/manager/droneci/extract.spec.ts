import { readFileSync } from 'fs';
import { resolve } from 'path';

const { extractPackageFile } = require('../../../lib/manager/droneci/extract');

const droneYAML = readFileSync(
  resolve('test/manager/droneci/_fixtures/.drone.yml'),
  'utf8'
);

describe('lib/manager/droneci/extract', () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty', () => {
      expect(extractPackageFile('nothing here')).toBeNull();
    });

    it('extracts multiple image lines', () => {
      const res = extractPackageFile(droneYAML);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(4);
    });
  });
});
