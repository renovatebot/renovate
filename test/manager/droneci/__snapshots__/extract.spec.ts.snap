// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/manager/droneci/extract extractPackageFile() extracts multiple image lines 1`] = `
Array [
  Object {
    "currentDigest": undefined,
    "currentValue": "1.8.1-alpine",
    "datasource": "docker",
    "depName": "elixir",
    "depType": "docker",
    "managerData": Object {
      "lineNumber": 5,
    },
  },
  Object {
    "commitMessageTopic": "Node.js",
    "currentDigest": "sha256:36adc17e9cceab32179d3314da9cb9c737ffb11f0de4e688f407ad6d9ca32201",
    "currentValue": "10.0.0",
    "datasource": "docker",
    "depName": "amd64/node",
    "depType": "docker",
    "managerData": Object {
      "lineNumber": 16,
    },
  },
  Object {
    "currentDigest": undefined,
    "currentValue": "5.7.24",
    "datasource": "docker",
    "depName": "mysql",
    "depType": "docker",
    "managerData": Object {
      "lineNumber": 23,
    },
  },
  Object {
    "currentDigest": undefined,
    "currentValue": "alpine",
    "datasource": "docker",
    "depName": "redis",
    "depType": "docker",
    "managerData": Object {
      "lineNumber": 28,
    },
  },
]
`;
